scrapy/spidermiddlewares/offsite.py

```diff
@@ -50,7 +51,15 @@ async def process_spider_output_async(
     def _filter(self, request: Any, spider: Spider) -> bool:
         if not isinstance(request, Request):
             return True
-        if request.dont_filter or self.should_follow(request, spider):
+        if request.dont_filter:
+            warnings.warn(
+                "The dont_filter filter flag is deprecated in OffsiteMiddleware. "
+                "Set 'allow_offsite' to True in Request.meta instead.",
+                ScrapyDeprecationWarning,
+                stacklevel=2,
+            )
+            return True
```

Comment:
Hm, I wonder if it'd be better to just keep supporting dont_filter, without a deprecation warning.

First, the flag says "dont_filter", and we respect it, and don't filter out the request. An argument can be made though that "don't filter" only applies to deduplication filter, not to other types of filtering. I think that's valid, Bbut that's not the current flag behavior, and also it's not in the flag name (it's not "dont_deduplicate").

Second, the user doesn't control all the don't_filter flags, Scrapy and other components can be setting this flag. For example, the default start_requests implementation uses dont_filter=True. 

It seems it's not possibile to "deprecate don_filter flag in OffsiteMiddleware", because the user might be setting this flag not for the OffsiteMiddleware, but for other Scrapy components, but the request may still end up in OffsiteMiddleware.

Reply:
Makes sense. In that case, we need to clarify that the `dont_filter` value also affects some middlewares as well and not just for the Scheduler.